File: Interview Cake: Apple Stocks/java/GetMaxProfit.java
// https://www.interviewcake.com/question/java/stock-price

//   int[] stockPricesYesterday = new int[]{10, 7, 5, 8, 11, 9};
//   getMaxProfit(stockPricesYesterday);
//   returns 6 (buying for $5 and selling for $11)

import java.util.Vector;
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

public class GetMaxProfit {
    public static void main(String []args) {
        
        Integer[] prices = new Integer[]{11, 10, 9, 8, 7, 5};
        
        int maxProfit = Integer.MIN_VALUE;
        
        for (int i = 0; i < prices.length; ++i) {
            int beforePrice = prices[i];
            for (int j = i + 1; j < prices.length; ++j) {
                int afterPrice = prices[j];
                if (i < j) {
                    int profit = afterPrice - beforePrice;
                    maxProfit = Math.max(maxProfit, profit);
                }
            }
        }
        
        System.out.println(maxProfit);
        
    }
}


// This is the correct answer from interviewcake
/*
  public int getMaxProfit(int[] stockPricesYesterday) {

    // make sure we have at least 2 prices
    if (stockPricesYesterday.length < 2) {
        throw new IllegalArgumentException("Getting a profit requires at least 2 prices");
    }

    // we'll greedily update minPrice and maxProfit, so we initialize
    // them to the first price and the first possible profit
    int minPrice = stockPricesYesterday[0];
    int maxProfit = stockPricesYesterday[1] - stockPricesYesterday[0];

    // start at the second (index 1) time
    // we can't sell at the first time, since we must buy first,
    // and we can't buy and sell at the same time!
    // if we started at index 0, we'd try to buy /and/ sell at time 0.
    // this would give a profit of 0, which is a problem if our
    // maxProfit is supposed to be /negative/--we'd return 0!
    for (int i = 1; i < stockPricesYesterday.length; i++) {
        int currentPrice = stockPricesYesterday[i];

        // see what our profit would be if we bought at the
        // min price and sold at the current price
        int potentialProfit = currentPrice - minPrice;

        // update maxProfit if we can do better
        maxProfit = Math.max(maxProfit, potentialProfit);

        // update minPrice so it's always
        // the lowest price we've seen so far
        minPrice = Math.min(minPrice, currentPrice);
    }

    return maxProfit;
}
*/

File: interview_questions/get_max_profit/c++/GetMaxProfit.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <algorithm>


void populateVector(std::vector<int>& p);

int main() {
    
    std::vector<int> prices;
    populateVector(prices);

    // for (int val : prices) {
    //     cout << val << endl;
    // }
        
    
    int minPrice = prices[0];
    int maximumProfit = prices[1] - prices[0];


    std::vector<int>::iterator it;
    for (it = ++prices.begin() ; it != prices.end() ; ++it) {
        int currentPrice = *it;
        
        // Check what the profit could be if we bought at min price
        // and sold at current price
        int potentialProfit = currentPrice - minPrice;
        
        // Check which profit is max
        maximumProfit = std::max(maximumProfit, potentialProfit);
        
        // Update min price
        minPrice = std::min(minPrice, currentPrice);
    
    }
    
    std::cout << maximumProfit << "\n";
    
    
    int i = 10;
    // std::cout << "Pre increment" << std::endl;
    // std::cout << ++i<< std::endl; //Pat: 11
    // std::cout << i << std::endl; // Pat: 11
    // i = 10;
    // std::cout << "Post increment" << std::endl;
    // std::cout << i++ << std::endl; //10
    // std::cout << i << std::endl; //11
    // i = 10;
    // std::cout << "Pre decrement" << std::endl;
    // std::cout << --i << std::endl; //9
    // std::cout << i << std::endl; //9
    // i = 10;
    // std::cout << "Post decrement" << std::endl;
    // std::cout << i-- << std::endl; //10
    // std::cout << i << std::endl; //9
    
    // i = 10;
    // std::cout << ++i + i++ << std::endl; // 11 + 12
    // std::cout << i << std::endl; // 23
    
    i = 10;
    std::cout << --i + i++ << std::endl;
    std::cout << i << std::endl;
    
    // i = 10;
    // std::cout << i-- + --i << std::endl;
    // std::cout << i << std::endl;
    
    
}

void populateVector(std::vector<int>& p) {
    p.push_back(12);
    p.push_back(11);
    p.push_back(10);
    p.push_back(9);
    p.push_back(8);
    p.push_back(7);
}